mongo/integration/unified/unified_spec_test.go

```diff
@@ -0,0 +1,252 @@
+	err = CheckSchemaVersion(testFile.SchemaVersion)
+	assert.Nil(t, err, "schema version %q not supported: %v", testFile.SchemaVersion, err)
+
+	// Create mtest wrapper, which will skip the test if needed.
+	mtOpts := mtest.NewOptions().
+		RunOn(testFile.RunOnRequirements...).
+		CreateClient(false)
+	mt := mtest.New(t, mtOpts)
+	defer mt.Close()
+
+	for _, testCase := range testFile.TestCases {
+		mtOpts := mtest.NewOptions().
+			RunOn(testCase.RunOnRequirements...).
+			CreateClient(false)
+		mt.RunOpts(testCase.Description, mtOpts, func(mt *mtest.T) {
+			runTestCase(mt, testFile, testCase)
+		})
+	}
+}
+
+func runTestCase(mt *mtest.T, testFile TestFile, testCase *TestCase) {
+	if testCase.SkipReason != nil {
+		mt.Skipf("skipping for reason: %q", *testCase.SkipReason)
+	}
+	if _, ok := skippedTestDescriptions[testCase.Description]; ok {
+		mt.Skip("skipping due to known failure")
+	}
+
+	testCtx := NewTestContext(mtest.Background)
+
+	defer func() {
+		// If anything fails while doing test cleanup, we only log the error because the actual test may have already
+		// failed and that failure should be preserved.
+
+		for _, err := range DisableUntargetedFailPoints(testCtx) {
+			mt.Log(err)
+		}
+		for _, err := range DisableTargetedFailPoints(testCtx) {
+			mt.Log(err)
+		}
+		for _, err := range Entities(testCtx).Close(testCtx) {
+			mt.Log(err)
+		}
+		// Tests that failed or started a transaction should terminate any sessions left open on the server. This is
+		// reqiured even if the test attempted to commit/abort the transaction because an abortTransaction command can
+		// fail if it's sent to a mongos that isn't aware of the transaction.
+		if mt.Failed() || testCase.StartsTransaction() {
+			if err := TerminateOpenSessions(mtest.Background); err != nil {
+				mt.Logf("error terminating open transactions after failed test: %v", err)
+			}
+		}
+	}()
+
+	// Set up collections based on the file-level initialData field.
+	for _, collData := range testFile.InitialData {
+		if err := collData.CreateCollection(testCtx); err != nil {
+			mt.Fatalf("error setting up collection %q: %v", collData.Namespace(), err)
+		}
+	}
+
+	// Set up entities based on the file-level createEntities field. For client entities, if the test will configure
+	// a fail point, set a low heartbeatFrequencyMS value into the URI options map if one is not already present.
+	// This speeds up recovery time for the client if the fail point forces the server to return a state change
+	// error.
+	shouldSetHeartbeatFrequency := testCase.SetsFailPoint()
+	for idx, entity := range testFile.CreateEntities {
+		for entityType, entityOptions := range entity {
+			if shouldSetHeartbeatFrequency && entityType == "client" {
+				if entityOptions.URIOptions == nil {
+					entityOptions.URIOptions = make(bson.M)
+				}
+				if _, ok := entityOptions.URIOptions["heartbeatFrequencyMS"]; !ok {
+					entityOptions.URIOptions["heartbeatFrequencyMS"] = lowHeartbeatFrequency
```

Comment:
Yeah, that appears to be a contradiction with the spec. Do you think it's worth filing a ticket for this? Also, based on reading the code, it seems like we will honor the 500ms minimum for <= 4.2 (4.4+ doesn't necessarily honor it due to streamable).

Reply:
I think it might be worth it. If a user was configuring with `heartbeatFrequencyMS=1` before and running against a 4.2 server, then upgrades, that could result in them hammering their servers with awaitable ismaster commands that respond immediately since `maxAwaitTimeMS` in awaitable ismaster is set to `heartbeatFrequencyMS`. I filed https://jira.mongodb.org/browse/GODRIVER-1778.